lnbits/db.py

```diff
@@ -47,27 +49,67 @@ def _parse_timestamp(value, _):
             lambda value, curs: float(value) if value is not None else None,
         )
     )
-    register_type(
-        new_type(
-            (1082, 1083, 1266),
-            "DATE2INT",
-            lambda value, curs: time.mktime(value.timetuple())
-            if value is not None
-            else None,
-        )
-    )
 
     register_type(new_type((1184, 1114), "TIMESTAMP2INT", _parse_timestamp))
 else:
     if os.path.isdir(settings.lnbits_data_folder):
         DB_TYPE = SQLITE
+        sqlite3.register_converter("json", lambda val: json.loads(val.decode("utf-8")))
+        sqlite3.register_converter("timestamp", lambda val: int(val))
     else:
         raise NotADirectoryError(
             f"LNBITS_DATA_FOLDER named {settings.lnbits_data_folder} was not created"
             f" - please 'mkdir {settings.lnbits_data_folder}' and try again"
         )
 
 
+class QueryValues(list):
+    """
+    A small wrapper around `list` to help with building SQL queries.
+    It makes sure the order of placeholders in the query and passed values match
+    by appending each value when the placeholder inserted.
+
+    It is especially useful in cases like `json_path`,
+    which produces database-varying sql containing a placeholder with an according value.
+    Without this helper, the function would have to inject the value into the query string directly, which is a security risk.
+
+    For example usage, see tests/core/test_db.py
+    """
+
+    def json_path(self, colname: str, *path: str, type_: Type = None):
+        if DB_TYPE == SQLITE:
+            path_value = f"$.{'.'.join(path)}"
+            accessor = f"json_extract({colname}, ?)"
+        else:
+            # https://www.postgresql.org/docs/9.3/functions-json.html
+            path_value = "{" + ",".join(path) + "}"
+            accessor = f"{colname} #>> ?"
+            if type_ == int:
+                accessor = f"({accessor})::int"
+        self.append(path_value)
+        return accessor
+
+    def json_partial_update(self, colname: str, value: dict):
+        self.append(value)
+        if DB_TYPE == SQLITE:
+            return f"{colname} = json_patch({colname}, ?)"
+        else:
+            return f"{colname} = {colname}::jsonb || ?"
+
+    def placeholder(self, value: Any):
```

Comment:
- can this method be "private" (`_placeholder`)?
- is it ever called outside this class?

Reply:
It is not called, but I thought it can be leaved if anyone wanted to be specific about a placeholder being inserted. The `__call__` wraps it for convenience. But it could be private yes